### ISSUE TITLE
client not reconnect when server is on after a shut down 

### ISSUE THREAD

**Issue by datka134** [binaryminds/react-native-sse#37]
seem like the reconnect not working or i miss some thing , pls help me

**Comment by EmilJunker** [comment]
What kind of shut down are you taking about? Did the server shut down and restart, and you are expecting the client to reconnect afterwards? Or did the client temporarily lose its connection to the internet? Was the client shut down? Please clarify what you mean.

Honestly, I don't see any reason why the SSE connection should automatically continue after any one such incident. I don't think there is anything wrong with the current behavior.

**Comment by Kief5555** [comment]
I'm having the exact same issue - when the connection is closed, it starts a new request despite me calling `se.close()`

**Comment by EmilJunker** [comment]
@Kief5555 Which version of react-native-sse are you using?
Can you please check if the response text from the SSE endpoint you are connecting to contains `reconnect:` values. That would maybe explain why the client automatically reconnects to the server after the connection is closed.

**Comment by Kief5555** [comment]
@EmilJunker Latest version. I've actually fixed this problem by setting pollingInterval to 0 and clearing event listeners on close.

**Comment by EmilJunker** [comment]
@Kief5555 Thank you for the prompt response and for sharing the workaround. However, that shouldn't be necessary. The latest version is supposed to have a mechanism that prevents it from even trying to reconnect once you've called `sse.close()`.

**Comment by Kief5555** [comment]
@EmilJunker I will double check. I installed today so I assumed it was latest..

**Comment by EmilJunker** [comment]
@Kief5555 If you can recreate this behavior with the latest version, it would be great if you could share some debug logs so we can get to the bottom of this.

**Comment by Kief5555** [comment]
> @Kief5555 If you can recreate this behavior with the latest version, it would be great if you could share some debug logs so we can get to the bottom of this.

I can confirm I'm on 1.2.0. How can I provide debug logs? Perhaps a URL to my event stream?

**Comment by EmilJunker** [comment]
@Kief5555 Thank you for checking and for your cooperation in helping to fix this unintended behavior.

If the event stream you are connecting to is publicly available and doesn't contain anything secret, you can share the URL. This would certainly be very helpful. 

Other than that, you can pass `debug: true` in the EventSource options parameter to get logs from react-native-sse. That would give us some insight into what is going on under the hood.

**Comment by EmilJunker** [comment]
@Kief5555 I really don't want to bother you and I understand if you are busy with other things, but if you can still reproduce this behavior it would be greatly appreciated if you could produce some logs (with option `debug: true`) and post them here so we can investigate the problem 👀

**Comment by Kief5555** [comment]
@EmilJunker Oops. Didn't see your previous message. I'm so sorry. The link is private, I can share via mail or discord if you would like.

**Comment by EmilJunker** [comment]
> Oops. Didn't see your previous message. I'm so sorry. The link is private, I can share via mail or discord if you would like.

@Kief5555 No worries, this isn't urgent. You can send the URL to a8bfbe9842@outlook.com and I'll have a look.

**Comment by Kief5555** [comment]
> > Oops. Didn't see your previous message. I'm so sorry. The link is private, I can share via mail or discord if you would like.
> 
> 
> 
> @Kief5555 No worries, this isn't urgent. You can send the URL to a8bfbe9842@outlook.com and I'll have a look. 

Sent.

**Comment by EmilJunker** [comment]
@Kief5555 Sorry for not getting back to this sooner, I only just found the time to check the URL you sent me. I tried to reproduce the problem, but my findings are a bit unclear tbh.

This is how I set up the SSE connection:

```ts
const sseEndpoint = '<YOUR_URL>';

const mySse = new EventSource(sseEndpoint, {
    debug: true,
});

const openHandler = (e: EventSourceEvent<'open'>) => {
    console.log('SSE Open', e);
};

const messageHandler = (e: EventSourceEvent<'message'>) => {
    console.log('SSE Message', e);
};

const errorHandler = (e: EventSourceEvent<'error'>) => {
    console.log('SSE Error', e);
};

const closeHandler = (e: EventSourceEvent<'close'>) => {
    console.log('SSE Close', e);
};

mySse.addEventListener('open', openHandler);
mySse.addEventListener('message', messageHandler);
mySse.addEventListener('error', errorHandler);
mySse.addEventListener('close', closeHandler);
```

I have found that if I call sse.close() within the first four to five seconds after the connection is opened, everything works normally.

Here are the logs I am getting in this case (the lines starting with DEBUG are internal logs coming from react-native-sse). As you can see, there is no more activity after I closed the connection:

```
 DEBUG  [EventSource] Will open new connection in 500 ms.
 DEBUG  [EventSource][onreadystatechange] ReadyState: HEADERS_RECEIVED(2), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 LOG  SSE Open {"type": "open"}
 DEBUG  [EventSource][onreadystatechange][OPEN] Connection opened.
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 LOG  SSE Message {"data": "{\"status\":true,\"message\":\"1 person is currently signing\",\"progress\":5}", "lastEventId": null, "type": "message", "url": "..."}
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 LOG  SSE Message {"data": "{\"status\":true,\"message\":\"Retrieving Information\",\"progress\":10}", "lastEventId": null, "type": "message", "url": "..."}
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 LOG  SSE Message {"data": "{\"status\":true,\"message\":\"Validating Certificate\",\"progress\":14}", "lastEventId": null, "type": "message", "url": "..."}
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 LOG  SSE Message {"data": "{\"status\":true,\"message\":\"Checking Device Registration\",\"progress\":19}", "lastEventId": null, "type": "message", "url": "..."}
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 LOG  SSE Message {"data": "{\"status\":true,\"message\":\"Account ID: 136\",\"progress\":19}", "lastEventId": null, "type": "message", "url": "..."}
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 LOG  SSE Message {"data": "{\"status\":true,\"message\":\"Retrieving Certificate\",\"progress\":23}", "lastEventId": null, "type": "message", "url": "..."}
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 LOG  SSE Message {"data": "{\"status\":true,\"message\":\"Retrieving Mobileprovision\",\"progress\":23}", "lastEventId": null, "type": "message", "url": "..."}
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 LOG  SSE Message {"data": "{\"status\":true,\"message\":\"Checking Device Status\",\"progress\":41}", "lastEventId": null, "type": "message", "url": "..."}
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 LOG  SSE Message {"data": "{\"status\":true,\"message\":\"Checking for errors\",\"progress\":41}", "lastEventId": null, "type": "message", "url": "..."}
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 DEBUG  [EventSource][onreadystatechange] ReadyState: LOADING(3), status: 200
 LOG  SSE Close {"type": "close"}
```

However, if I let the connection open for more than five seconds, it becomes impossible to close it. I think the reason for this is that there is so much SSE activity that the whole app just freezes and I can't do anything. Even simple UI state refreshes don't work (I tried to increment a counter on the press of a button, but it didn't react while the SSE was open).

In other words, the problem is not that the SSE connection doesn't close when sse.close() is called. The problem is that sse.close() doesn't even get called at all because the app process is so busy with the event handling.

PS: I'm actually not sure if the problem I am experiencing is the same one you are talking about since you said you were able to fix it by setting pollingInterval to 0 and clearing the event handlers. But I am able to reproduce it pretty reliably, so I'm assuming this is it.

**Comment by Kief5555** [comment]
Hey @EmilJunker, 

I see. The issue with me is that even calling .close() didn't do anything, tho the react UI updates just fine and the app doesn't have a hard time.

I will try to get reproducible code for you as soon as I have time.

**Comment by Kief5555** [comment]
@EmilJunker, I got it for you.
```
			const es = new EventSource(url, {
				// pollingInterval: 0
				debug: true
			});


			es.addEventListener('open', () => {
				//do something
			});
			interface Event {
				data: string | null;
			}
			es.addEventListener('message', async (event: Event) => {
				const data = JSON.parse(event.data as string);
				if (!data.status) {
					es.close();
					throw new Error(data.message);
				}
			});

			es.addEventListener('close', () => {
				es.close();
				// Remove only the specific event listeners related to this signingID
				es.removeAllEventListeners();
			});
```


Logs (`debug: true`): 
[message.txt](https://github.com/binaryminds/react-native-sse/files/14303806/message.txt)

**Comment by EmilJunker** [comment]
@Kief5555 Thank you very much for the code and the logs. I now understand much better what is happening. I have three remarks:

1. The fact that the connection automatically restarts after it is closed by the server (i.e. where it says `[DONE] Operation done.` in the logs) is the expected behaviour. If you want to disable this retry operation, you need to set pollingInterval to 0 (as you already figured out by yourself). Also removing the event handlers is not necessary.

2. There currently exists no logic to inform the client when the server has closed the connection. As written above, when the server closes the connection, the client just automatically reopens it (unless you set pollingInterval to 0). Crucially, no `close` event gets emitted in this case. That's why the `es.close()` call that you make in the listener for the `close` event never actually gets called.

3. The only time the `close` event gets emitted is when the client closes the connection, i.e. by calling `es.close()`. That's why it doesn't actually make sense to call `es.close()` in the close handler. In fact, doing so resulted in an infinite call loop for me and eventually lead to a `Maximum call stack size exceeded` error. This is something that we should probably fix, but unrelated to this issue.

**Comment by Kief5555** [comment]
> @Kief5555 Thank you very much for the code and the logs. I now understand much better what is happening. I have three remarks:
> 
> 
> 
> 1. The fact that the connection automatically restarts after it is closed by the server (i.e. where it says `[DONE] Operation done.` in the logs) is the expected behaviour. If you want to disable this retry operation, you need to set pollingInterval to 0 (as you already figured out by yourself). Also removing the event handlers is not necessary.
> 
> 
> 
> 2. There currently exists no logic to inform the client when the server has closed the connection. As written above, when the server closes the connection, the client just automatically reopens it (unless you set pollingInterval to 0). Crucially, no `close` event gets emitted in this case. That's why the `es.close()` call that you make in the listener for the `close` event never actually gets called.
> 
> 
> 
> 3. The only time the `close` event gets emitted is when the client closes the connection, i.e. by calling `es.close()`. That's why it doesn't actually make sense to call `es.close()` in the close handler. When the close handler is triggered, it means `es.close()` has already been called. In fact, calling it again in the close handler resulted in an infinite call loop for me and eventually lead to a `Maximum call stack size exceeded` error 😅. This is something that we should probably fix, but unrelated to this issue.

Ah, I see now. I feel like the expected behaviour is to close when the server closes a connection. Thank you very much for helping me report the issue!

**Comment by EmilJunker** [comment]
> Ah, I see now. I feel like the expected behaviour is to close when the server closes a connection. Thank you very much for helping me report the issue!

I agree that this behavour is a bit unintuitive, but the SSE stadard mandates those automatic reconnections. However, if the server responds with HTTP status 204, then the client should actually not attempt to reconnect.

According to https://javascript.info/server-sent-events:
`If the server wants the browser to stop reconnecting, it should respond with HTTP status 204.`

I don't think react-native-sse currently checks the HTTP status before reconnecting, so this is another thing that we should probably fix.

**Comment by Kief5555** [comment]
@EmilJunker, I'm also noticing another issue,

With the event handler, when I `throw new Error` it seems to ignore all try-catch blocks and just returns an uncaught error. This is causing my application to crash (even tho I have a try catch to prevent the problem and retry).

**Comment by EmilJunker** [comment]
@Kief5555 Thank you for bringing this up. I'll investigate this problem tomorrow if I have time. In the meanwhile, please open a new issue for this and provide some code to reproduce the problem.

**Comment by Kief5555** [comment]
> > With the event handler, when I `throw new Error` it seems to ignore all try-catch blocks and just returns an uncaught error. This is causing my application to crash (even tho I have a try catch to prevent the problem and retry). Even adding the error event handler doesn't seem to catch it
> 
> @Kief5555 Thank you for bringing this up. I'll investigate this problem tomorrow if I have time. In the meanwhile, please open a new issue for this and provide some code to reproduce the problem.

Hi @EmilJunker, Sounds good. I will just quickly confirm tomorrow thats an issue on my side again before opening another issue. (I'm currently busy so please hold!)
